### PR TITLE
4.x : Allow nodes to hold other child nodes

### PR DESCRIPTION
**Pull request by deepakrkris**
implement composite pattern with parent-child

### CODE REVIEW

File: test/unit/graph/composite.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+'use strict';
+
+const Composite = require('../../../lib/datamodel/graph').Composite;
+const expect = require('../../helpers/expect');
+
+describe('Graph : Composite', function() {
+  describe('contains()', function() {
+    let phase, middleware;
+
+    before(function() {
```
Comment (superkhau) [inline test/unit/graph/composite.js]: name your anonymouse function and move down under the it's

File: test/unit/graph/composite.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+          this.contains(Middleware);
+        }
+      }
+      middleware = new Middleware({}, 'model1', 'foo');
+      phase = new Phase({}, 'model2', 'bar');
+    });
+
+    it('adds placeholder for child nodes', function() {
+      expect(phase.children).to.have.property('Middleware');
+    });
```
Comment (superkhau) [inline test/unit/graph/composite.js]: add empty line

File: test/unit/graph/composite.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    describe('add()', function() {
+      it('adds child node to parent', function() {
```
Comment (superkhau) [inline test/unit/graph/composite.js]: adds the child node to the parent node

File: test/unit/graph/composite.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        constructor(graph, domain, id) {
+          super(graph, domain, id);
+          this.contains(Middleware);
+        }
+      }
+      middleware = new Middleware({}, 'model1', 'foo');
+      phase = new Phase({}, 'model2', 'bar');
+    });
+
+    it('adds placeholder for child nodes', function() {
```
Comment (superkhau) [inline test/unit/graph/composite.js]: adds a placeholder for child noes

**Comment by deepakrkris** [comment]
@superkhau fixed comments

File: test/unit/graph/composite.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    describe('add()', function() {
+      it('adds a child node to the parent', function() {
+        phase.add(middleware);
+        expect(phase.children.Middleware).to.have.property('foo');
+      });
+
+      it('parent.child(childName) gets child', function() {
+        expect(phase.middleware('foo')).to.eql(middleware);
+      });
+    });
```
Comment (superkhau) [inline test/unit/graph/composite.js]: empty line here

**Pull request merged**